### PR TITLE
New TF examples

### PR DESCRIPTION
**Pull request by Rocketknight1**
Opening a PR to get some feedback on the new TF example style before I write the rest.

Don't merge it yet, I haven't even finalized the filenames!

### CODE REVIEW

File: examples/text-classification/new_tf_text_classification.py [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+from transformers.utils import check_min_version
+
+logger = logging.getLogger(__name__)
+
+# region Helper classes
+class DataSequence(tf.keras.utils.Sequence):
+    # We use a Sequence object to load the data. Although it's completely possible to load your data as Numpy/TF arrays
+    # and pass those straight to the Model, this constrains you in a couple of ways. Most notably, it requires all
+    # the data to be padded to the length of the longest input example, and it also requires the whole dataset to be
+    # loaded into memory. If these aren't major problems for you, you can skip the sequence object in your own code!
```
Comment (LysandreJik) [inline examples/text-classification/new_tf_text_classification.py:57]: Great comment!

File: examples/text-classification/new_tf_text_classification.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+    def __init__(self, dataset, non_label_column_names, batch_size, labels, shuffle=True):
+        super().__init__()
+        # Retain all of the columns not present in the original data - these are the ones added by the tokenizer
+        self.data = {key: dataset[key] for key in dataset.features.keys()
+                     if key not in non_label_column_names
+                     and key != 'label'}
+        data_lengths = {len(array) for array in self.data.values()}
+        assert len(data_lengths) == 1  # Assert that all arrays have the same length
```
Comment (LysandreJik) [inline examples/text-classification/new_tf_text_classification.py:65]: As a general style observation, we prefer not to have empty asserts in the code base (applies below as well)
```suggestion
        assert len(data_lengths) == 1, "Error message"  # Assert that all arrays have the same length
```

File: examples/text-classification/new_tf_text_classification.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+    default_data_collator,
+    set_seed,
+)
+from transformers.trainer_utils import is_main_process, get_last_checkpoint
+from transformers.utils import check_min_version
+
+logger = logging.getLogger(__name__)
+
+# region Helper classes
+class DataSequence(tf.keras.utils.Sequence):
```
Comment (LysandreJik) [inline examples/text-classification/new_tf_text_classification.py:53]: Maybe too complex for this example, but I believe it's generally best to have the longest sequence be part of the initial batch, as the full batch then has the maximum (`seq_len x batch_size`) that will be encountered during training; if the training was to OOM at some point, it would do so right off the bat rather than 1hr in.

Unless I'm mistaken, that's not done here, would it be easy to add?

File: examples/text-classification/new_tf_text_classification.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+        default=None, metadata={"help": "Pretrained tokenizer name or path if not the same as model_name"}
+    )
+    cache_dir: Optional[str] = field(
+        default=None,
+        metadata={"help": "Where do you want to store the pretrained models downloaded from huggingface.co"},
+    )
+    use_fast_tokenizer: bool = field(
+        default=True,
+        metadata={"help": "Whether to use one of the fast tokenizer (backed by the tokenizers library) or not."},
+    )
```
Comment (LysandreJik) [inline examples/text-classification/new_tf_text_classification.py:207]: I think we should just always recommend a fast tokenizer, so I'm not even sure we really need that here. Thoughts @sgugger?

File: examples/text-classification/new_tf_text_classification.py [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+        learning_rate=training_args.learning_rate,
+        beta_1=training_args.adam_beta1,
+        beta_2=training_args.adam_beta2,
+        epsilon=training_args.adam_epsilon,
+        clipnorm=training_args.max_grad_norm)
+    if is_regression:
+        loss = tf.keras.losses.MeanSquaredError()
+    else:
+        loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+    model.compile(optimizer=optimizer, loss=loss)
```
Comment (LysandreJik) [inline examples/text-classification/new_tf_text_classification.py:371]: I wouldn't put this in `# region Load pretrained model and tokenizer`, but maybe something like `# region Define optimizer, loss and compilation`? I feel like this is something that should be separate

**Review by LysandreJik — COMMENTED** [review]
I find this wonderfully simple to get through, great job @Rocketknight1!

I left a few comments below.

File: examples/text-classification/new_tf_text_classification.py (hunk 4 above)
Comment (sgugger) [reply examples/text-classification/new_tf_text_classification.py:207]: Yes, this predates v4 when we had to actually set `use_fast=True`, we can probably remove that arg now.

File: examples/text-classification/new_tf_text_classification.py [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+        model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+    output_dir = Path(training_args.output_dir)
+    output_dir.mkdir(parents=True, exist_ok=True)
+    # endregion
+
+    # region Checkpoints
+    # Detecting last checkpoint.
+    checkpoint = None
+    if len(os.listdir(training_args.output_dir)) > 0 and not training_args.overwrite_output_dir:
+        if (output_dir / 'config.json').is_file() and (output_dir / 'tf_model.h5').is_file():
```
Comment (sgugger) [inline examples/text-classification/new_tf_text_classification.py:243]: Nit: we should import the constant containing the names from transformers.file_utils.

File: examples/text-classification/new_tf_text_classification.py [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+    if training_args.do_train:
+        for index in random.sample(range(len(train_dataset)), 3):
+            logger.info(f"Sample {index} of the training set: {train_dataset[index]}.")
+
+    if data_args.pad_to_max_length:
+        data_collator = default_data_collator
+    elif training_args.fp16:
+        data_collator = DataCollatorWithPadding(tokenizer, pad_to_multiple_of=8)
+    else:
+        data_collator = None
```
Comment (sgugger) [inline examples/text-classification/new_tf_text_classification.py:446]: This is PyTorch specific and should be removed.

File: examples/text-classification/new_tf_text_classification.py [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+        # if last_checkpoint is not None:
+        #     checkpoint = last_checkpoint
+        # elif os.path.isdir(model_args.model_name_or_path):
+        #     # Check the config from that potential checkpoint has the right number of labels before using it as a
+        #     # checkpoint.
+        #     if AutoConfig.from_pretrained(model_args.model_name_or_path).num_labels == num_labels:
+        #         checkpoint = model_args.model_name_or_path
+        #
+        # if last_checkpoint is not None:
+        #     model.load_weights(last_checkpoint)
```
Comment (sgugger) [inline examples/text-classification/new_tf_text_classification.py:461]: I think we can clean this up for now.

**Review by sgugger — APPROVED** [review]
Looking great! Thanks a lot for working on this!

File: examples/text-classification/new_tf_text_classification.py (hunk 1 above)
Comment (Rocketknight1) [reply examples/text-classification/new_tf_text_classification.py:57]: Thanks!

File: examples/text-classification/new_tf_text_classification.py [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+
+    def on_epoch_end(self):
+        # If we're shuffling, reshuffle the data order after each epoch
+        if self.shuffle:
+            self.permutation = np.random.permutation(self.data_length)
+
+    def __getitem__(self, item):
+        # Note that this yields a batch, not a single sample
+        batch_start = item * self.batch_size
+        batch_end = (item + 1) * self.batch_size
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/text-classification/new_tf_text_classification.py:89]: We don't need it, though we could put it in for clarity! If you pass slice indices past the end of a list/array, you just get a slice up to the end of the list/array, e.g.
```
>>> mylist = [1, 2, 3, 4, 5]
>>> mylist[3:8]
[4, 5]
>>>
```

File: examples/text-classification/new_tf_text_classification.py (hunk 8 above)
Comment (Rocketknight1) [reply examples/text-classification/new_tf_text_classification.py:461]: Done!

File: examples/text-classification/new_tf_text_classification.py (hunk 7 above)
Comment (Rocketknight1) [reply examples/text-classification/new_tf_text_classification.py]: Done!

File: examples/text-classification/new_tf_text_classification.py [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+import tensorflow as tf
+from datasets import load_dataset, load_metric
+
+import transformers
+from transformers import (
+    AutoConfig,
+    TFAutoModelForSequenceClassification,
+    AutoTokenizer,
+    DataCollatorWithPadding,
+    EvalPrediction,
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/text-classification/new_tf_text_classification.py]: Done!

File: examples/text-classification/new_tf_text_classification.py (hunk 4 above)
Comment (Rocketknight1) [reply examples/text-classification/new_tf_text_classification.py:208]: Done!

File: examples/text-classification/new_tf_text_classification.py (hunk 6 above)
Comment (Rocketknight1) [reply examples/text-classification/new_tf_text_classification.py:244]: Done!

File: examples/text-classification/new_tf_text_classification.py (hunk 5 above)
Comment (Rocketknight1) [reply examples/text-classification/new_tf_text_classification.py:369]: Done!

File: examples/tensorflow/text-classification/README.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Text classification examples
+
+This folder contains some scripts showing examples of *text classification* with the Transformers library.
```
Comment (sgugger) [inline examples/tensorflow/text-classification/README.md:19]:
```suggestion
This folder contains some scripts showing examples of *text classification* with the 🤗 Transformers library.
```

File: examples/tensorflow/text-classification/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,3 @@
+datasets >= 1.1.3
+sentencepiece != 0.1.92
+protobuf
```
Comment (sgugger) [inline examples/tensorflow/text-classification/requirements.txt:3]: Add tensorflow here :-)

**Review by sgugger — COMMENTED** [review]
I like the new README!

**Review by sgugger — APPROVED** [review]
LGTM, left a few comments, mostly about formatting, and I think we could add accuracy as a metric if we don't have a regression task.

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 13]
```diff
@@ -0,0 +1,537 @@
...
+    into argparse arguments to be able to specify them on
+    the command line.
+    """
+
+    train_file: Optional[str] = field(default=None,
+                                      metadata={"help": "A csv or a json file containing the training data."})
+    validation_file: Optional[str] = field(default=None,
+                                           metadata={"help": "A csv or a json file containing the validation data."})
+    test_file: Optional[str] = field(default=None,
+                                     metadata={"help": "A csv or a json file containing the test data."})
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py]: I forgot the formatting run again, sorry. Fixed!

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+        datefmt="%m/%d/%Y %H:%M:%S",
+        handlers=[logging.StreamHandler(sys.stdout)],
+    )
+    logger.setLevel(logging.INFO)
+
+    # Set the verbosity to info of the Transformers logger (on main process only):
+    if is_main_process(training_args.local_rank):
+        transformers.utils.logging.set_verbosity_info()
+        transformers.utils.logging.enable_default_handler()
+        transformers.utils.logging.enable_explicit_format()
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py:283]: Done!

File: examples/tensorflow/text-classification/requirements.txt (hunk 12 above)
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/requirements.txt:3]: Done!

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 15]
```diff
@@ -0,0 +1,537 @@
...
+    # a single grammatical sentence, when the task requires it.
+    #
+    # If the CSVs/JSONs contain only one non-label column, the script does single sentence classification on this
+    # single column. You can easily tweak this behavior (see below)
+    #
+    # In distributed training, the load_dataset function guarantee that only one local process can concurrently
+    # download the dataset.
+    data_files = {"train": data_args.train_file,
+                  "validation": data_args.validation_file,
+                  "test": data_args.test_file}
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py]: Done!

File: examples/tensorflow/text-classification/run_text_classification.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py]: Done!

File: examples/tensorflow/text-classification/run_text_classification.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py]: Done!

File: examples/tensorflow/text-classification/requirements.txt (hunk 12 above)
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/requirements.txt:3]: Done!

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 16]
```diff
@@ -0,0 +1,537 @@
...
+        beta_2=training_args.adam_beta2,
+        epsilon=training_args.adam_epsilon,
+        clipnorm=training_args.max_grad_norm,
+    )
+    if is_regression:
+        # TODO Make sure we handle saving/loading regression models properly
+        loss = tf.keras.losses.MeanSquaredError()
+    else:
+        loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+    model.compile(optimizer=optimizer, loss=loss)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py:389]: Done!

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 17]
```diff
@@ -0,0 +1,537 @@
...
+            eval_dataset = None
+
+        callbacks = [SavePretrainedCallback(output_dir=training_args.output_dir)]
+        model.fit(training_dataset, validation_data=eval_dataset, epochs=10, callbacks=callbacks)
+    elif "validation" in datasets:
+        # If there's a validation dataset but no training set, just evaluate the metrics
+        eval_dataset = DataSequence(
+            eval_dataset, non_label_column_names, batch_size=training_args.per_device_eval_batch_size, labels=True
+        )
+        print("Computing loss on validation data...")
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py]: Done!

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 18]
```diff
@@ -0,0 +1,532 @@
...
+            eval_dataset = None
+
+        callbacks = [SavePretrainedCallback(output_dir=training_args.output_dir)]
+        model.fit(training_dataset, validation_data=eval_dataset, epochs=10, callbacks=callbacks)
+    elif "validation" in datasets:
+        # If there's a validation dataset but no training set, just evaluate the metrics
+        eval_dataset = DataSequence(
+            eval_dataset, non_label_column_names, batch_size=training_args.per_device_eval_batch_size, labels=True
+        )
+        logger.info("Computing metrics  on validation data...")
```
Comment (sgugger) [inline examples/tensorflow/text-classification/run_text_classification.py:500]:
```suggestion
        logger.info("Computing metrics on validation data...")
```

**Review by sgugger — APPROVED** [review]
Great work!

**Review by LysandreJik — APPROVED** [review]
LGTM, great work @Rocketknight1!

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 19]
```diff
@@ -0,0 +1,532 @@
...
+        )
+        if "validation" in datasets:
+            eval_dataset = DataSequence(
+                eval_dataset, non_label_column_names, batch_size=training_args.per_device_eval_batch_size, labels=True
+            )
+        else:
+            eval_dataset = None
+
+        callbacks = [SavePretrainedCallback(output_dir=training_args.output_dir)]
+        model.fit(training_dataset, validation_data=eval_dataset, epochs=10, callbacks=callbacks)
```
Comment (philschmid) [inline examples/tensorflow/text-classification/run_text_classification.py:494]: Is it on purpose that `epochs=10` is hardcoded? shouldn't this be `epochs=training_args. num_train_epochs`?

**Review by philschmid — COMMENTED** [review]
Hey just took a look at your the `TF` example it looks amazing. I noticed one thing what seemed strange to me

File: examples/tensorflow/text-classification/run_text_classification.py (hunk 19 above)
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py:494]: Aagh, you're totally right. Fixing it!

**Comment by Rocketknight1** [comment]
I opened a new branch and PR at #11360 to avoid dealing with rebasing after the folder structure was changed around
